hazelcast/src/test/java/com/hazelcast/cache/JCacheTypeChecks.java

```diff
@@ -0,0 +1,372 @@
+            }
+        });
+
+        Cache rawCache = (Cache) cache;
+        rawCache.put(1, 1);
+
+        Class<?> clazz = Class.forName("com.hazelcast.cache.impl.AbstractCacheProxyBase");
+        Field configField = clazz.getDeclaredField("cacheConfig");
+        configField.setAccessible(true);
+
+        CacheConfig cacheConfig = (CacheConfig) configField.get(rawCache);
+        cacheConfig.setTypes(Integer.class, String.class);
+
+        cache.getAndRemove(1);
+    }
+
+    @Test(expected = ClassCastException.class)
+    public void test_check_types_on_replace() throws Exception {
+        final String cacheName = randomMapName();
+
+        CacheManager cacheManager = cachingProvider1.getCacheManager();
+        assertNotNull(cacheManager);
+
+        assertNull(cacheManager.getCache(cacheName));
+
+        CompleteConfiguration<Integer, String> config = new CacheConfig<Integer, String>();
+
+        Cache<Integer, String> cache = cacheManager.createCache(cacheName, config);
+        assertNotNull(cache);
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                CacheManager cm2 = cachingProvider2.getCacheManager();
+                assertNotNull(cm2.getCache(cacheName));
+            }
+        });
+
+        Cache rawCache = (Cache) cache;
+        rawCache.put(1, 1);
+
+        Class<?> clazz = Class.forName("com.hazelcast.cache.impl.AbstractCacheProxyBase");
+        Field configField = clazz.getDeclaredField("cacheConfig");
+        configField.setAccessible(true);
+
+        CacheConfig cacheConfig = (CacheConfig) configField.get(rawCache);
+        cacheConfig.setTypes(Integer.class, String.class);
+
+        rawCache.replace(1, 3);
+    }
+
+    @Test(expected = ClassCastException.class)
+    public void test_check_types_on_getandreplace() throws Exception {
+        final String cacheName = randomMapName();
+
+        CacheManager cacheManager = cachingProvider1.getCacheManager();
+        assertNotNull(cacheManager);
+
+        assertNull(cacheManager.getCache(cacheName));
+
+        CompleteConfiguration<Integer, String> config = new CacheConfig<Integer, String>();
+
+        Cache<Integer, String> cache = cacheManager.createCache(cacheName, config);
+        assertNotNull(cache);
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                CacheManager cm2 = cachingProvider2.getCacheManager();
+                assertNotNull(cm2.getCache(cacheName));
+            }
+        });
+
+        Cache rawCache = (Cache) cache;
+        rawCache.put(1, 1);
+
+        Class<?> clazz = Class.forName("com.hazelcast.cache.impl.AbstractCacheProxyBase");
+        Field configField = clazz.getDeclaredField("cacheConfig");
+        configField.setAccessible(true);
+
+        CacheConfig cacheConfig = (CacheConfig) configField.get(rawCache);
```

Comment:
here there are some duplications, it could be good if you write a method like `buildConfig`with some parameters, so readability can be increased. and every test can call that method.